File: claude-usage-panel@fschmutz.github.io/lib/cost.js
// Optional cost layer: the official usage API does not expose the dollar cost
// of token usage on subscription plans, so we shell out to `ccusage` (which
// computes it from the local ~/.claude/projects/*.jsonl logs).
//
// Only an installed `ccusage` runs. There is no `npx ccusage@latest`
// fallback: that fetched and ran the newest unpinned npm release on every
// poll, inside the session that holds the Claude OAuth token.

import {run} from './proc.js';

export const CCUSAGE_ARGV = ['ccusage', 'blocks', '--active', '--json'];

/**
 * Run `ccusage blocks --active --json` and resolve the active block cost.
 * @returns {Promise<{costUSD: number, tokens: number} | null>}
 */
export async function fetchActiveCost() {
    const {ok, stdout} = await run(CCUSAGE_ARGV);
    if (!ok || !stdout)
        return null;
    let block;
    try {
        block = JSON.parse(stdout)?.blocks?.[0];
    } catch {
        return null;
    }
    if (!block)
        return null;
    return {
        costUSD: Number(block.costUSD) || 0,
        tokens: Number(block.totalTokens) || 0,
    };
}

File: macos/Sources/ClaudeUsagePanel/Cost.swift
import Foundation

// Optional cost layer: the official usage API does not expose dollar cost on
// subscription plans, so we shell out to `ccusage` (computed from the local
// ~/.claude/projects/*.jsonl logs). Mirrors the GNOME extension's lib/cost.js.
// Only an installed `ccusage` runs: no `npx ccusage@latest` fallback, which
// fetched and ran the newest unpinned npm release on every poll.

struct ActiveCost {
    let costUSD: Double
    let tokens: Int
}

enum Cost {
    static let argv = ["ccusage", "blocks", "--active", "--json"]

    /// Run `ccusage blocks --active --json`. Returns nil if unavailable.
    static func fetchActiveCost() async -> ActiveCost? {
        run(argv)
    }

    private static func run(_ argv: [String]) -> ActiveCost? {
        let r = Shell.run("/usr/bin/env", argv, env: Shell.toolEnvironment)
        guard r.ok,
            let json = try? JSONSerialization.jsonObject(with: Data(r.out.utf8)) as? [String: Any],
            let blocks = json["blocks"] as? [[String: Any]],
            let first = blocks.first
        else { return nil }
        return ActiveCost(
            costUSD: (first["costUSD"] as? NSNumber)?.doubleValue ?? 0,
            tokens: (first["totalTokens"] as? NSNumber)?.intValue ?? 0)
    }
}
